File: site/non-migrated/lib/mtext-shorten.php
<?php
# @(#) $Id$

require_once('lib/xml.php');
require_once('lib/charsets.php');
require_once('lib/text.php');

function strpos_all($haystack, $needle, &$found) {
    if (is_array($needle)) {
        foreach ($needle as $s)
            strpos_all($haystack, $s, $found);
    } else {
        $pos = strpos($haystack, $needle);
        while ($pos !== false) {
            $found[] = $pos;
            $pos = strpos($haystack, $needle, $pos + 1);
        }
    }
    return count($found);
}

function getShortenLength($s, $len, $mdlen, $pdlen) {
    debugLog(LL_FUNCTIONS, 'getShortenLength(s=%,len=%,mdlen=%,pdlen=%)',
            array($s, $len, $mdlen, $pdlen));
    if (strlen($s) <= $len + $pdlen) {
        $return = strlen($s);
        debugLog(LL_FUNCTIONS, 'getShortenLength() returned %',
                array($return));
        return $return;
    }
    $st = $len - $mdlen;
    $st = $st < 0 ? 0 : $st;
    $c = substr($s, $st, $mdlen + $pdlen);
    $patterns = array("\x1F",
                      array('. ', '! ', '? ', ".\n", "!\n", "?\n"),
                      array(': ', ', ', '; ', ') ', ":\n", ",\n", ";\n", ")\n"));
    foreach ($patterns as $pat) {
        $matches = array();
        if (!strpos_all($c, $pat, $matches))
            continue;
        $bestpos = -1;
        foreach ($matches as $pos) {
            if ($bestpos < 0 || abs($bestpos - $mdlen) > abs($pos - $mdlen))
                $bestpos = $pos;
        }
        $matchLen = is_array($pat) ? 2 : 1;
        $return = $bestpos + $st + $matchLen;
        debugLog(LL_FUNCTIONS, 'getShortenLength() returned %', array($return));
        return $return;
    }
    debugLog(LL_FUNCTIONS, 'getShortenLength() returned %', array($len));
    return $len;
}

function hasMarkup($s) {
    for ($i = 0; $i < strlen($s); $i++) {
        if (strpos("<>&=~_^[]{}'", $s[$i]) !== false)
            return true;
        if ($s[$i] == '/' && ($i == 0 || $s[$i - 1] == ' ' || $s[$i - 1] == ':'))
            return true;
    }
    return false;
}

function cleanLength($s) {
    if (!hasMarkup($s))
        return strlen($s);
    $xml = new MTextToLineXML();
    $xml->parse(convertToXMLText($s));
    $xml->free();
    return strlen($xml->getLine());
}

function shortenUniversal($s, $len, $mdlen, $pdlen, $clearTags = false,
                          $suffix = '') {
    debugLog(LL_FUNCTIONS,
             'shortenUniversal(s=%,len=%,mdlen=%,pdlen=%,clearTags=%,suffix=%)',
             array($s, $len, $mdlen, $pdlen, $clearTags, $suffix));
    $hasMarkup = hasMarkup($s);
    if ($hasMarkup) {
        $xml = new MTextToLineXML();
        $xmlText = convertToXMLText($s);
        debugLog(LL_DETAILS, 'convertToXMLText(s)=%', array($xmlText));
        $xml->parse($xmlText);
        $xml->free();
        $line = $xml->getLine();
    } else {
        $line = $s;
    }
    $n = getShortenLength($line, $len, $mdlen, $pdlen);
    $c = $n >= strlen($line) ? '' : $suffix;
    if ($hasMarkup) {
        $xml1 = new MTextShortenXML($n, $clearTags);
        $xml1->parse(convertToXMLText($s));
        $xml1->free();
        $return = $xml1->getShort().$c;
    } else {
        $return = substr($s, 0, $n).$c;
    }
    debugLog(LL_FUNCTIONS, 'shortenUniversal() returned %', array($return));
    return delicateAmps($return);
}

function shorten($s, $len, $mdlen, $pdlen) {
    return shortenUniversal($s, $len, $mdlen, $pdlen);
}

function shortenNote($s, $len, $mdlen, $pdlen) {
    return shortenUniversal($s, $len, $mdlen, $pdlen, true, '...');
}
?>

File: site/src/main/java/ua/org/migdal/mtext/MtextShorten.java
package ua.org.migdal.mtext;

import java.util.ArrayList;
import java.util.List;

import ua.org.migdal.util.XmlConverter;
import ua.org.migdal.util.XmlUtils;
import ua.org.migdal.util.exception.XmlConverterException;

class MtextShorten {

    private static final String[][] ENDS = {
            {"\u001F"},
            {". ", "! ", "? ", ".\n", "!\n", "?\n"},
            {": ", ", ", "; ", ") ", ":\n", ",\n", ";\n", ")\n"}
    };

    private static int indexOfAll(String haystack, String needle, List<Integer> found) {
        int pos = haystack.indexOf(needle);
        while (pos >= 0) {
            found.add(pos);
            pos = haystack.indexOf(needle, pos + 1);
        }
        return found.size();
    }

    private static int indexOfAll(String haystack, String[] needles, List<Integer> found) {
        for (String needle : needles) {
            indexOfAll(haystack, needle, found);
        }
        return found.size();
    }

    private static int getShortenLength(String s, int len, int mdlen, int pdlen) {
        if (s.length() <= len + pdlen) {
            return s.length();
        }
        int st = len - mdlen;
        st = st < 0 ? 0 : st;
        String c = s.substring(st, mdlen + pdlen);
        for (String[] patterns : ENDS) {
            List<Integer> matches = new ArrayList<>();
            if (indexOfAll(c, patterns, matches) <= 0) {
                continue;
            }
            int bestpos = -1;
            for (int pos : matches) {
                if (bestpos < 0 || Math.abs(bestpos - mdlen) > Math.abs(pos - mdlen)) {
                    bestpos = pos;
                }
            }
            int matchLen = patterns.length > 1 ? 2 : 1;
            return bestpos + st + matchLen;
        }
        return len;
    }

    private static int cleanLength(String s) throws XmlConverterException {
        if (!XmlUtils.hasMarkup(s)) {
            return s.length();
        }
        MtextToLine handler = new MtextToLine();
        XmlConverter.convert(s, handler);
        return handler.getLine().length();
    }

    public static String shorten(String s, int len, int mdlen, int pdlen) {
        return shorten(s, len, mdlen, pdlen, false, "");
    }

    public static String shorten(String s, int len, int mdlen, int pdlen, boolean clearTags, String suffix) {
        boolean hasMarkup = XmlUtils.hasMarkup(s);

        try {
            String line;
            if (hasMarkup) {
                MtextToLine handler = new MtextToLine();
                XmlConverter.convert(s, handler);
                line = handler.getLine().toString();
            } else {
                line = s;
            }

            int n = getShortenLength(line, len, mdlen, pdlen);
            String c = n >= line.length() ? "" : suffix;

            String result;
            if (hasMarkup) {
                MtextToLength handler = new MtextToLength(n, clearTags);
                XmlConverter.convert(s, handler);
                result = handler.getShortened().toString();
            } else {
                result = s.substring(0, n);
            }
            return XmlUtils.delicateAmps(result) + c;
        } catch (XmlConverterException e) {
            return String.format("<b>** %s: %s **</b>", e.getMessage(), e.getCause().getMessage());
        }
    }

}